### PR TITLE
feat(auto-battle): 支持多声音模板闪避识别

### PR DESCRIPTION
**Pull request by ImLaoBJie**
## 为什么修改

原有声音闪避识别只支持一个普通黄光/红光模板（`template_1.wav`），无法区分紫光（[控制技]）和 X 黄光（连携招架）提示音。本次扩展为三个声音模板（后续可继续扩展添加），同时移除仅为少量音频处理而引入的较重依赖，并保持识别结果与实时计算性能与旧实现相同。

## 改动要点

- 新增 `AudioTemplateEnum`，分别表示：
  - `template_1.wav` -> `NORMAL_DODGE`：普通黄光、红光闪避声音。
  - `template_2.wav` -> `PURPLE_DODGE`：紫光闪避声音。
  - `template_3.wav` -> `X_DODGE`：X 黄光闪避声音。
- 新增 `template_2.wav`、`template_3.wav` ，保留 `template_1.wav` 。
- 三个模板先分别计算相关系数；只要最高分超过阈值，就返回最高分对应的模板编号，未命中仍返回 `False`。
- 只有 `NORMAL_DODGE` 和 `X_DODGE` 会写入声音闪避状态；`PURPLE_DODGE` 只完成分类，不触发闪避操作，这是因为控制技提示出现后一定时长才BOSS会发动攻击。
- 模板初始化时一次性完成读取、单声道转换、重采样、高通滤波、标准化和 FFT 缓存，避免后续计算。
- 实时识别时只对录音做一次 FFT，再利用二维数组广播同时计算三个模板的相关系数，避免逐模板重复处理录音。
- 不同长度模板只在批量 FFT 时补零；用有效区间掩码排除补零区，并按各模板真实长度计算分母，保持与旧算法一致。
- 为录音缓冲区增加加锁复制，避免录音线程写入时识别线程同时读取同一数组。
- 静音或近似静音波形在标准差接近零时返回全零，避免出现 `NaN`。
- 保留声音匹配类型，可后续扩展。

## 等价实现与依赖调整

- 用 `numpy.mean(..., axis=1)` 等价替代 `librosa.to_mono` 的双声道逐帧平均。
- 用 `scipy.io.wavfile` 读取 WAV，并统一处理无符号整数、有符号整数和浮点数据类型。
- 用 `scipy.signal.resample_poly` 等价完成模板重采样。
- 用除以标准差等价替代 `sklearn.preprocessing.scale(..., with_mean=False)`。
- 从项目依赖和打包模块清单中移除 `librosa`、`scikit-learn` 及已无其他用途的传递依赖。
- 将源码实际直接使用的 `scipy==1.17.1` 和 `requests==2.34.2` 改为直接依赖；两者在旧锁文件中已经存在，因此没有新增安装包。
- 同步更新 `pyproject.toml`、`uv.lock`、`deploy/module_manifest.py` 和虚拟环境。
- 将玩家 skill 中闪避助手的实现说明从 `librosa` 改为 `scipy`。

### 依赖变化

| 类型 | 依赖 |
|---|---|
| 移除的直接依赖 | `librosa` |
| 不再需要的锁文件包 | `audioread`、`joblib`、`lazy-loader`、`librosa`、`llvmlite`、`msgpack`、`narwhals`、`numba`、`pooch`、`scikit-learn`、`soundfile`、`soxr`、`threadpoolctl` |
| 补充为直接依赖 | `scipy`、`requests` |
| 新增锁文件包 | 无 |

按旧 `uv.lock` 中 Windows x64、CPython 3.11 对应 wheel 的压缩大小统计，上述 13 个不再需要的包合计约 **52.77 MiB（55.33 MB）**。这是安装下载包的可复核口径，实际解压后的虚拟环境占用通常更大。

## 多模板计算与性能

三个模板并不是在每个实时窗口中依次完整运行三遍旧算法：模板的滤波、标准化和 FFT 都在初始化阶段缓存，实时阶段只对录音滤波、标准化并做一次 FFT，然后通过二维数组广播同时完成三个模板的频域乘法和逆 FFT。

批量实现与逐模板旧算法逐项对比，误差不超过 `1e-12`，可认为和原实现一致。

### 测试条件

- 副本：恶名狩猎 / 基塔布鲁。
- 运行方式：体力刷本，连续执行 5 次。
- 采样方式：旁路音频脚本持续记录新三模板批量算法和旧单模板算法的完整浮点结果，不参与游戏操作；计划完成次数增加时记录分轮边界。
- 采样结果：共 14,101 个音频窗口；第 1～5 轮分别结束于 126.797、200.719、271.391、340.922、441.657 秒。

### 性能采样

采样的逐窗口耗时如下：

CPU 和内存使用同一批 256 个真实录音窗口分别重放测量，并与上面的逐窗口耗时分开执行，避免内存跟踪干扰耗时结果。CPU 使用当前采样线程的线程时间；“单线程 CPU”表示隔离重放期间该采样线程的占用率，不是整个游戏或一条龙进程的总 CPU 占用。内存同时记录 Python/NumPy 可跟踪临时分配峰值和进程 RSS（驻留内存）峰值。

| 算法 | 256 窗口 CPU 时间 | 平均 CPU 时间/窗口 | 单线程 CPU | 临时分配峰值增量 | 进程 RSS 峰值 | RSS 峰值增量 |
|---|---:|---:|---:|---:|---:|---:|
| 新算法 | 250.00 ms | 0.977 ms | 85.37% | 2.078 MiB | 190.781 MiB | 0.004 MiB |
| 旧算法 | 218.75 ms | 0.854 ms | 83.39% | 1.519 MiB | 190.781 MiB | 0.000 MiB |

在这批窗口中，新算法线程 CPU 时间增加 31.25 ms，约 **0.122 ms/窗口（14.3%）**；可跟踪临时分配峰值增加约 **0.559 MiB**。两者的进程 RSS 峰值在 0.001 MiB 显示精度下相同，说明新增三个模板的实时计算没有造成可见的进程常驻内存增长。RSS 是同一采样进程的整体值，不能解释为算法独占内存；算法临时内存差异以“临时分配峰值增量”为主要对比口径。

#### 五轮合计与均值

| 模板 | 超过阈值总次数 | 作为最大值总次数 | 每轮超过阈值均值 | 每轮作为最大值均值 |
|---|---:|---:|---:|---:|
| `NORMAL_DODGE` | 119 | 119 | 23.8 | 23.8 |
| `PURPLE_DODGE` | 7 | 7 | 1.4 | 1.4 |
| `X_DODGE` | 26 | 26 | 5.2 | 5.2 |
| 旧算法 | 119 | 119 | 23.8 | 23.8 |

本次总共记录 152 个命中窗口，没有任何窗口出现两个模板同时超过阈值，所以两列计数恰好相同。`NORMAL_DODGE` 与旧模板五轮均为 119 次，说明批量实现完整保留了旧模板的识别结果。新增的 26 个 X 命中窗口聚成 5 段连续声音，每轮各 1 段；这些窗口的旧模板分数均未超过阈值，属于旧实现没有的新识别结果。

#### 命中相关系数分布

下表只统计本次五轮中已经超过当前阈值 0.1、并最终分类为对应模板的窗口。Q1 和 Q3 分别是第 25、75 百分位，可作为箱线图的箱体上下边界；旧模板统计其自身超过 0.1 的窗口。

| 模板 | 样本数 | 最小值 | Q1 | 中位数 | 平均值 | Q3 | 最大值 |
|---|---:|---:|---:|---:|---:|---:|---:|
| `NORMAL_DODGE` | 119 | 0.100005 | 0.109289 | 0.128299 | 0.138016 | 0.156894 | 0.237119 |
| `PURPLE_DODGE` | 7 | 0.100558 | 0.101354 | 0.108470 | 0.114508 | 0.117648 | 0.154521 |
| `X_DODGE` | 26 | 0.100692 | 0.115612 | 0.128741 | 0.164234 | 0.173067 | 0.493293 |
| 旧算法 | 119 | 0.100005 | 0.109289 | 0.128299 | 0.138016 | 0.156894 | 0.237119 |
 
#### 不同阈值下的命中保留数

因为本次保存了所有超过 0.1 的命中窗口，所以可以准确回算把统一阈值提高后的保留数量；无法用这批数据估计阈值低于 0.1 时会增加多少命中。判断条件仍为严格大于候选阈值。

| 候选阈值 | `NORMAL_DODGE` | `PURPLE_DODGE` | `X_DODGE` | 旧模板 |
|---:|---:|---:|---:|---:|
| 0.10 (当前) | 119 | 7 | 26 | 119 |
| 0.11 | 86 | 3 | 21 | 86 |
| 0.12 | 69 | 2 | 17 | 69 |
| 0.13 | 56 | 1 | 12 | 56 |
| 0.15 | 39 | 1 | 9 | 39 |
| 0.20 | 7 | 0 | 5 | 7 |

#### 五轮合计与均值

| 模板 | 超过阈值总次数 | 作为最大值总次数 | 每轮超过阈值均值 | 每轮作为最大值均值 |
|---|---:|---:|---:|---:|
| `NORMAL_DODGE` | 150 | 150 | 30.0 | 30.0 |
| `PURPLE_DODGE` | 0 | 0 | 0.0 | 0.0 |
| `X_DODGE` | 35 | 35 | 7.0 | 7.0 |
| 旧模板 | 150 | 150 | 30.0 | 30.0 |

本次共记录 185 个命中窗口，没有任何窗口出现两个模板同时超过 0.1。`NORMAL_DODGE` 与旧模板都是 150 次，逐窗口最大绝对误差为 `8.33e-17`，说明背景音乐存在时批量实现仍保留旧单模板结果。35 个 X 命中窗口按 1 秒间隔聚成 5 段，每轮各 1 段；这些窗口均是旧单模板没有覆盖的新增结果。`PURPLE_DODGE` 本次为 0 次，说明本轮背景音乐没有形成超过 0.1 的紫光误识别。

与无背景音乐的五轮结果相比，本次普通、X 和紫光计数分别由 119、26、7 变为 150、35、0。加入背景杂音后，误识率理所当然上升，但不同模板之间的区分度更好。

#### 命中相关系数分布

| 模板 | 样本数 | 最小值 | Q1 | 中位数 | 平均值 | Q3 | 最大值 |
|---|---:|---:|---:|---:|---:|---:|---:|
| `NORMAL_DODGE` | 150 | 0.100169 | 0.114420 | 0.127423 | 0.137010 | 0.151698 | 0.282884 |
| `PURPLE_DODGE` | 0 | - | - | - | - | - | - |
| `X_DODGE` | 35 | 0.103512 | 0.115554 | 0.132716 | 0.156069 | 0.163884 | 0.361014 |
| 旧模板 | 150 | 0.100169 | 0.114420 | 0.127423 | 0.137010 | 0.151698 | 0.282884 |

#### 不同阈值下的命中保留数

| 候选阈值 | `NORMAL_DODGE` | `PURPLE_DODGE` | `X_DODGE` | 旧模板 |
|---:|---:|---:|---:|---:|
| 0.10 (当前) | 150 | 0 | 35 | 150 |
| 0.11 | 122 | 0 | 31 | 122 |
| 0.12 | 96 | 0 | 22 | 96 |
| 0.13 | 71 | 0 | 18 | 71 |
| 0.15 | 38 | 0 | 10 | 38 |
| 0.20 | 10 | 0 | 6 | 10 |

### 阈值是否调整

- 建议
`self.trigger_threshold: float = 0.1 # 触发阈值`
调整为
`self.trigger_threshold: float = 0.11 # 触发阈值`
但可能减少闪避的识别率，因此本轮未做调整

- 模板原文件均为 48 kHz 双声道，运行时统一转换为 32 kHz 单声道，转换确实改变波形并丢弃超高频信息，但当前识别结果证明影响可忽略。

### 一次 X 音效为什么会识别出多个窗口

五轮战斗中实际每轮只出现一次 X 黄光音效，但分别记录到 2、6、7、6、5 个 `X_DODGE` 命中窗口。录音器维护最近 0.5 秒的滑动缓冲区，每 10 ms 写入一批新录音，识别逻辑约每 20 ms 检查一次。一次 X 音效持续时间会跨过多次检查，而且声音内部可能有多个与模板相似的片段，所以同一声音会在连续窗口中多次超过阈值。

旧算法也是类似效果，但`check_dodge_audio`已对识别轮询间隔作出限制，每次轮询最多触发一次，相邻间隔触发概率较低，影响不大

### `紫光[控制技]` 音效误识别原因

未开启背景音乐前， 7 个窗口识别为紫光，其中 5 个分数只有 0.1006～0.1120，刚刚越过统一阈值。战斗中的短促高频音效可能与紫光模板的局部波形相似，因此会产生少量误识别；`argmax` 只负责从三个分数中选最大值，并不能表示“未知声音”。这是预期可接受的误识别。

### 声音模板来源

直接从游戏中录制并将声音导出。

实际上`ZenlessZoneZero_Data\StreamingAssets\Audio\Windows\Full\...`可能存放了闪避的原始音效，使用该音效作为模板可能更精确。
但存放的声音文件均以数字编号命名，[AnimeWwise](https://github.com/Escartem/AnimeWwise)和[ZZAR](https://github.com/Pucas01/ZZAR)等项目给出了角色对话和文件名称的映射，但没有闪避等游戏音效的名称映射，使用匹配算法也无法找到（高匹配度的听起来完全不是），要想找出实际的闪避原声，要人工听几万个声音文件，这是不可能的。
如果后续有大能能找出来，再做替换。

### 其他

- **由于该改动涉及闪避识别等重要功能，建议尽量多测试再考虑合并。**

### CODE REVIEW

**Review by idk500 — COMMENTED** [review]
去 librosa/sklearn 改纯 scipy 实现是好事，但相关系数分母和标准化实现都重写了，数值行为与旧算法有差异，阈值 0.1 直接沿用——是否有与旧算法的实测对比（命中/误报率）或测试仓用例？template_2/3 的来源与采样率兼容性（不同位深 WAV）是否实测覆盖？没有回归依据的话建议补一组实测对比再合。

**Comment by ImLaoBJie** [comment]
> 去 librosa/sklearn 改纯 scipy 实现是好事，但相关系数分母和标准化实现都重写了，数值行为与旧算法有差异，阈值 0.1 直接沿用——是否有与旧算法的实测对比（命中/误报率）或测试仓用例？template_2/3 的来源与采样率兼容性（不同位深 WAV）是否实测覆盖？没有回归依据的话建议补一组实测对比再合。

@idk500 
- 一致性结果见 PR 的「多模板计算与性能」：与scipy.signal.correlate一致的原实现([legacy_get_max_corr](https://github.com/ImLaoBJie/zzz-od-test/blob/audio_detect_upgrade/test/zzz_od/auto_battle/auto_battle_dodge_context/manual_live_benchmark.py#L104))和新实现([AutoBattleDodgeContext.get_max_corr](https://github.com/ImLaoBJie/zzz-od-test/blob/audio_detect_upgrade/test/zzz_od/auto_battle/auto_battle_dodge_context/manual_live_benchmark.py#L100))对比，绝对误差不超过 1e-12。实战对比见 PR 的「严谨的五轮实战验证」及「开启背景音乐后的五轮实战验证」。两组各 5 轮：旧实现和新实现在相同阈值下的命中数都是相同的；开启背景音乐时新旧实现的最大绝对误差为 8.33e-17。

- 测试用例： [manual_live_benchmark.py](https://github.com/ImLaoBJie/zzz-od-test/blob/audio_detect_upgrade/test/zzz_od/auto_battle/auto_battle_dodge_context/manual_live_benchmark.py)，包含数值等价和实战采样用例。

- 实时录音按 32 kHz 直接从系统采集，不管是提供好的模板（template_2/3录制方式与template_1一致, 48 kHz）还是用户自己换，都会一律resample至 32 kHz，不存在兼容性问题。
